### PR TITLE
add squared hinge loss

### PR DESCRIPTION
**Pull request by Patrick1860**
@hectorgon @deerzq 
Took a shot to add the squared hinge loss to learn Scala. First Scala PR ever so PTAL and let me know if there is anything terribly wrong. Thanks

### CODE REVIEW

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 1]
```diff
@@ -212,11 +214,17 @@ object FullRankLinearTrainer {
               val scores = model.scoreFlatFeature(flatFeatures)
               val posScore = scores.values(posIdx)
               val negScore = scores.values(negIdx)
-              val loss = (posMargin - negMargin) - (negScore - posScore)
+              val loss = (posMargin - negMargin) + (negScore - posScore)
```
Comment (Patrick1860) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: I think it should be margin + negScore - posScore unless I miss anything.

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 2]
```diff
@@ -212,11 +214,17 @@ object FullRankLinearTrainer {
...
               if (loss > 0.0) {
                 val grad = new FloatVector(dim)
-                grad.values(posIdx) = -1.0f
-                grad.values(negIdx) = 1.0f
+                if (lossType == "l1" ) {
+                  grad.values(posIdx) = -1.0f
+                  grad.values(negIdx) = 1.0f
+                } else {
+                  grad.values(posIdx) = (-2.0 * loss).toFloat
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: If you define loss as 0.5 * || x || ^2 then the gradient is x

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala (hunk 1 above)
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: Yes I think so too.. I had a typo!

File: training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala [hunk 3]
```diff
@@ -212,11 +214,17 @@ object FullRankLinearTrainer {
...
+              val loss = (posMargin - negMargin) + (negScore - posScore)
               if (loss > 0.0) {
                 val grad = new FloatVector(dim)
-                grad.values(posIdx) = -1.0f
-                grad.values(negIdx) = 1.0f
+                if (lossType == "l1" ) {
+                  grad.values(posIdx) = -1.0f
+                  grad.values(negIdx) = 1.0f
+                } else {
+                  grad.values(posIdx) = loss.toFloat
```
Comment (hectorgon) [inline training/src/main/scala/com/airbnb/aerosolve/training/FullRankLinearTrainer.scala]: -loss.toFloat

**Comment by hectorgon** [comment]
:+1:
